### PR TITLE
Restrict metadata keys to only support known keys

### PR DESCRIPTION
**Pull request by michelpmcdonald**
## Summary

- Lock down benefits Intake custom metadata keys by adding a config file that ties specific consumer with their allowed custom metadata keys.  Before this change, unsupported keys were allowed, now, unsupported keys will cause an error
- We need to allow specific consumers special metadata, for example, internal va consumers are allowed specific keys
- Team Banana peels, our code

## Related issue(s)
https://vajira.max.gov/browse/API-24188

## Testing done

unit tests added

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
Benefits Intake request consumer provided metadata
## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

**Review by ryan-mcneil — APPROVED** [review]
LGTM. ping me if you need an additional review after fixing the merge conflict 👌

File: lib/central_mail/utilities.rb [hunk 1]
```diff
@@ -3,12 +3,18 @@
...
     META_PART_NAME = 'metadata'
     DOC_PART_NAME = 'content'
     SUBMIT_DOC_PART_NAME = 'document'
     VALID_LOB = { 'CMP' => 'CMP', 'PMC' => 'PMC', 'INS' => 'INS', 'EDU' => 'EDU', 'VRE' => 'VRE', 'BVA' => 'BVA',
                   'FID' => 'FID', 'NCA' => 'NCA', 'OTH' => 'CMP' }.freeze
+
+    # Required Metadata fields
     REQUIRED_KEYS = %w[veteranFirstName veteranLastName fileNumber zipCode].freeze
+    # Supported fields that all consumers are permitted to provide
+    META_FIELD_NAMES_ALL_CONSUMERS = REQUIRED_KEYS + %w[source docType businessLine].freeze
```
Comment (kristen-brown) [inline lib/central_mail/utilities.rb:16]: Looks good! I was surprised to see that `source` is marked as "required" in the `DocumentUploadMetadata` section of our documentation. If it's not actually required by us or the upstream systems, we should probably update our documentation to match.

File: modules/vba_documents/lib/vba_documents/upload_validator.rb [hunk 2]
```diff
@@ -35,13 +35,16 @@ def validate_parts(model, parts)
...
+    def validate_metadata(metadata_input, submission_version:, consumer_name: '')
       metadata = JSON.parse(metadata_input)
       raise VBADocuments::UploadError.new(code: 'DOC102', detail: 'Invalid JSON object') unless metadata.is_a?(Hash)
 
       missing_keys = REQUIRED_KEYS - metadata.keys
+      extra_keys = metadata.keys - consumer_meta_field_names_allowed(consumer_name)
       if missing_keys.present?
         raise VBADocuments::UploadError.new(code: 'DOC102', detail: "Missing required keys: #{missing_keys.join(',')}")
+      elsif extra_keys.present?
+        raise VBADocuments::UploadError.new(code: 'DOC102', detail: "Invalid keys provided: #{extra_keys.join(',')}")
```
Comment (kristen-brown) [inline modules/vba_documents/lib/vba_documents/upload_validator.rb:47]: 👍

**Review by kristen-brown — CHANGES_REQUESTED** [review]
Left a couple of suggestions. Let me know if you have any questions!

I also thought we documented in the ticket that we wanted to use the `consumer_id` rather than the `consumer_name` (since it was less likely to change). Did that change?

File: lib/central_mail/utilities.rb [hunk 3]
```diff
@@ -18,6 +24,22 @@ module Utilities
...
+    def consumer_meta_field_names_allowed(consumer_name)
+      meta_fields_config_path = VBADocuments::Engine.root.join('config', 'metadata', 'internal_consumer_fields.yml')
+
+      consumer_metadata_fields = {}
+      if File.exist?(meta_fields_config_path)
+        consumer_metadata_fields = YAML.load_file(meta_fields_config_path) || {}
+      else
+        VBADocuments::Slack::Messager.new(
+          { warning: ":warning: #{self.class.name} consumer metadata fields config file does not exist",
+            recipient_file: recipient_file_path.to_s }
```
[the remark this answers is not in the document]
Comment (michelpmcdonald) [reply lib/central_mail/utilities.rb:36]: nice catch, thank you

File: lib/central_mail/utilities.rb (hunk 1 above)
Comment (michelpmcdonald) [reply lib/central_mail/utilities.rb:16]: Created ticket

File: modules/vba_documents/app/controllers/vba_documents/v2/uploads_controller.rb [hunk 4]
```diff
@@ -3,6 +3,7 @@
 require 'zip'
 require 'common/exceptions'
 require 'vba_documents/payload_manager'
+require 'vba_documents/pdf_inspector'
 require 'vba_documents/upload_validator'
 require './lib/webhooks/utilities'
```
[the remark this answers is not in the document]
Comment (michelpmcdonald) [reply modules/vba_documents/app/controllers/vba_documents/v2/uploads_controller.rb:8]: I did not realize we checked in both places, thank you

**Pull request closed without merging**